presto-main/src/main/java/io/prestosql/server/security/CertificateAuthenticator.java

```diff
@@ -30,32 +31,45 @@
 {
     private static final String X509_ATTRIBUTE = "javax.servlet.request.X509Certificate";
 
+    private final CertificateAuthenticatorManager authenticatorManager;
     private final UserMapping userMapping;
 
     @Inject
-    public CertificateAuthenticator(CertificateConfig config)
+    public CertificateAuthenticator(CertificateAuthenticatorManager authenticatorManager, CertificateConfig config)
     {
         requireNonNull(config, "config is null");
         this.userMapping = createUserMapping(config.getUserMappingPattern(), config.getUserMappingFile());
+        this.authenticatorManager = requireNonNull(authenticatorManager, "authenticatorManager is null");
+        authenticatorManager.setRequired();
     }
 
     @Override
     public Identity authenticate(HttpServletRequest request)
             throws AuthenticationException
     {
-        X509Certificate[] certs = (X509Certificate[]) request.getAttribute(X509_ATTRIBUTE);
-        if ((certs == null) || (certs.length == 0)) {
+        List<X509Certificate> certificates;
+        if (request.getAttribute(X509_ATTRIBUTE) == null) {
             throw new AuthenticationException(null);
         }
-        X500Principal principal = certs[0].getSubjectX500Principal();
+        else {
+            certificates = Arrays.asList((X509Certificate[]) request.getAttribute(X509_ATTRIBUTE));
```

Comment:
Use `ImmutableList.copyOf()`, since `asList()` is a mutable view -- changes to the list modify the array, and changes to the array are visible in the list. This would allow an authenticator to modify the original array in the servlet request by writing to the list. (not likely a practical concern, but in general we avoid allowing such things as it prevents mistakes)